commit 999b9c7a9cb1d7cbb8b4407ddfa33a2a39ee09b2
Author: Arne Fitzenreiter <arne.fitzenreiter@ipfire.org>
Date:   Wed Jun 6 08:02:10 2012 +0200

    bcm2835-ctl: fix alsamixer control.
    
    alsamixer read the volume for the screen controller so we had to
    scale the chipvol back to db for reading.

Index: linux-3.2.46/sound/arm/bcm2835-ctl.c
===================================================================
--- linux-3.2.46.orig/sound/arm/bcm2835-ctl.c	2013-07-26 19:32:35.000000000 +0000
+++ linux-3.2.46/sound/arm/bcm2835-ctl.c	2013-07-26 19:33:16.000000000 +0000
@@ -33,6 +33,19 @@
 
 #include "bcm2835.h"
 
+
+/* functions to convert alsa to chip volume and back. */
+int alsa2chip(int vol)
+{
+	return -((vol << 8) / 100);
+}
+
+int chip2alsa(int vol)
+{
+	return -((vol * 100) >> 8);
+}
+
+
 static int snd_bcm2835_ctl_info(struct snd_kcontrol *kcontrol,
 				struct snd_ctl_elem_info *uinfo)
 {
@@ -64,7 +77,7 @@
 	BUG_ON(!chip && !(chip->avail_substreams & AVAIL_SUBSTREAMS_MASK));
 
 	if (kcontrol->private_value == PCM_PLAYBACK_VOLUME)
-		ucontrol->value.integer.value[0] = chip->volume;
+		ucontrol->value.integer.value[0] = chip2alsa(chip->volume);
 	else if (kcontrol->private_value == PCM_PLAYBACK_MUTE)
 		ucontrol->value.integer.value[0] = chip->mute;
 	else if (kcontrol->private_value == PCM_PLAYBACK_DEVICE)
@@ -85,13 +98,10 @@
 			changed = 1;
 		}
 		if (changed
-		    || (ucontrol->value.integer.value[0] != chip->volume)) {
-			int atten;
+		    || (ucontrol->value.integer.value[0] != chip2alsa(chip->volume))) {
 
-			chip->volume = ucontrol->value.integer.value[0];
+			chip->volume = alsa2chip(ucontrol->value.integer.value[0]);
 			changed = 1;
-			atten = -((chip->volume << 8) / 100);
-			chip->volume = atten;
 		}
 
 	} else if (kcontrol->private_value == PCM_PLAYBACK_MUTE) {
